source/mesh_deformation/diffusion.cc

```diff
@@ -0,0 +1,587 @@
+      if (this->get_timestep() > 0.)
+        velocity /= this->get_timestep();
+      else
+        AssertThrow(false, ExcZero());
+
+      output = velocity;
+    }
+
+
+    template <int dim>
+    void Diffusion<dim>::check_diffusion_time_step (const DoFHandler<dim> &mesh_deformation_dof_handler,
+                                                    const std::set<types::boundary_id> &boundary_ids) const
+    {
+      // Initialize Gauss-Legendre quadrature for degree+1 quadrature points of the surface faces
+      const QGauss<dim-1> face_quadrature(mesh_deformation_dof_handler.get_fe().degree+1);
+
+      FEFaceValues<dim> fs_fe_face_values (this->get_mapping(), mesh_deformation_dof_handler.get_fe(), face_quadrature, update_default);
+
+      double min_local_conduction_timestep = std::numeric_limits<double>::max();
+
+      for (const auto &fscell : mesh_deformation_dof_handler.active_cell_iterators())
+        if (fscell->at_boundary() && fscell->is_locally_owned())
+          for (unsigned int face_no=0; face_no<GeometryInfo<dim>::faces_per_cell; ++face_no)
+            if (fscell->face(face_no)->at_boundary())
+              {
+                // Get the boundary indicator of current cell face
+                const types::boundary_id boundary_indicator
+                  = fscell->face(face_no)->boundary_id();
+
+                // Only consider the requested boundaries
+                if (boundary_ids.find(boundary_indicator) == boundary_ids.end())
+                  continue;
+
+                // Reinitialize update flags for current cell face
+                fs_fe_face_values.reinit (fscell, face_no);
+
+                // Calculate the corresponding conduction timestep
+                min_local_conduction_timestep = std::min(min_local_conduction_timestep,
+                                                         this->get_parameters().CFL_number*std::pow(fscell->face(face_no)->minimum_vertex_distance(),2.)
+                                                         / diffusivity);
+              }
+
+      // Get the global minimum timestep
+      const double min_conduction_timestep = Utilities::MPI::min (min_local_conduction_timestep, this->get_mpi_communicator());
+
+      double conduction_timestep = min_conduction_timestep;
+      if (this->convert_output_to_years())
+        conduction_timestep /= year_in_seconds;
+
+      AssertThrow (this->get_timestep() <= min_conduction_timestep,
+                   ExcMessage("The numerical timestep is too large for diffusion of the surface. Although the "
+                              "diffusion scheme is stable, note that the error increases linearly with the timestep. "
+                              "The diffusion timestep is: " + std::to_string(conduction_timestep) + ", while "
+                              "the advection timestep is: " + std::to_string (this->get_timestep ())));
+    }
+
+
+    /**
+     * A function that creates constraints for the velocity of certain mesh
```

Comment:
This comment should not be here, it should be in the header (but I think it is already there?)

Reply:
yes, removed the comment.